src/ocaml/parser/syntax.rs

```diff
@@ -0,0 +1,482 @@
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::Makeblock(pos, id, xs)))
+        }
+        "let" => {
+            let (i, bindings) = delimited(
+                pair(multispace1, tag("(")),
+                many1(preceded(multispace0, parse_let_binding)),
+                tag(")"),
+            )(i)?;
+            let (i, body) = preceded(multispace0, parse_syntax)(i)?;
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::Let(pos, bindings, body.into())))
+        }
+        "letrec" => {
+            let (i, bindings) = delimited(
+                pair(multispace1, tag("(")),
+                many1(preceded(multispace0, parse_letrec_binding)),
+                tag(")"),
+            )(i)?;
+            let (i, body) = preceded(multispace0, parse_syntax)(i)?;
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::Letrec(pos, bindings, body.into())))
+        }
+        "function" => {
+            let (i, args) = many0(preceded(multispace0, parse_function_arg))(i)?;
+            let (i, body) = preceded(
+                alt((
+                    delimited(multispace1, tag(": int"), multispace1),
+                    multispace1,
+                )),
+                parse_syntax,
+            )(i)?;
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::Function(pos, args, body.into())))
+        }
+        "apply" => {
+            let (i, func) = preceded(multispace1, parse_syntax)(i)?;
+            let (i, args) = many1(preceded(multispace0, parse_syntax))(i)?;
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::Apply(pos, func.into(), args)))
+        }
+        _ => {
+            // parse as a placeholder fallback s-expression
+            let (i, tail) = many0(preceded(
+                multispace0,
+                alt((parse_syntax, preceded(multispace0, parse_fallback_literal))),
+            ))(i)?;
+            let (i, _) = preceded(multispace0, tag(")"))(i)?;
+            let pos = Pos::from_upto(from, i);
+            Ok((i, LambdaSyntax::FallbackPrimitive(pos, head, tail)))
+        }
+    }
+}
+
+pub fn parse_syntax(from: Span<'_>) -> ParseResult<'_, LambdaSyntax> {
+    delimited(
+        multispace0,
+        alt((
+            parse_primitive_sexp,
+            parse_ident,
+            parse_numeric,
+            parse_string,
+            parse_char,
+            parse_record,
+        )),
+        multispace0,
+    )(from)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn set_pos(l: &LambdaSyntax, p: Pos) -> LambdaSyntax {
```

Comment:
It's cleaner if you implement equality by hand, ignoring the `Pos`, similar to how it was done for `Syntax`. You can also implement `Hash` following the same idea (it was also done for `Syntax`).

Reply:
D'oh, this is so much nicer! I don't even need this `set_pos` stuff at all